impl_core/src/main/java/io/opencensus/implcore/metrics/DoubleGaugeMetricImpl.java

```diff
@@ -0,0 +1,164 @@
+import javax.annotation.Nullable;
+
+/** Implementation of {@link DoubleGaugeMetric}. */
+public final class DoubleGaugeMetricImpl extends DoubleGaugeMetric implements Meter {
+  private final MetricDescriptor metricDescriptor;
+  private final ConcurrentMap<List<LabelValue>, Point> pointConcurrentHashMap =
+      new ConcurrentHashMap<List<LabelValue>, Point>();
+  private final List<LabelKey> labelKeys;
+
+  DoubleGaugeMetricImpl(String name, String description, String unit, List<LabelKey> labelKeys) {
+    this.metricDescriptor =
+        MetricDescriptor.create(name, description, unit, Type.GAUGE_DOUBLE, labelKeys);
+    this.labelKeys = labelKeys;
+  }
+
+  @Override
+  public Point addPoint(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  @Override
+  public <T> void addPoint(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    addAndGetPoint(labelValues, obj, checkNotNull(function, "function"));
+  }
+
+  @Override
+  public Point getDefaultPoint() {
+    List<LabelValue> labelValues = new ArrayList<LabelValue>(labelKeys.size());
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  private <T> Point addAndGetPoint(
+      List<LabelValue> labelValues, @Nullable T obj, @Nullable ToDoubleFunction<T> function) {
+    Point currentPoint = pointConcurrentHashMap.get(labelValues);
+    if (currentPoint != null) {
+      return currentPoint;
+    } else {
+      Point newPoint =
+          function != null
+              ? new PointImpl<T>(labelValues, obj, function)
+              : new PointImpl<T>(labelValues);
+
+      pointConcurrentHashMap.put(labelValues, newPoint);
+      return newPoint;
+    }
+  }
+
+  private List<TimeSeries> getTimeSeries(Clock clock) {
+    List<TimeSeries> timeSeriesList = new ArrayList<TimeSeries>(pointConcurrentHashMap.size());
+
+    for (Map.Entry<List<LabelValue>, Point> pointEntry : pointConcurrentHashMap.entrySet()) {
+      timeSeriesList.add(
+          TimeSeries.create(
+              pointEntry.getKey(),
+              Collections.singletonList(
+                  io.opencensus.metrics.export.Point.create(
+                      Value.doubleValue(((PointImpl) pointEntry.getValue()).get()), clock.now())),
+              null));
+    }
+    return timeSeriesList;
+  }
+
+  @Override
+  public Metric getMetric(Clock clock) {
+    return Metric.create(metricDescriptor, getTimeSeries(clock));
+  }
+
+  /** Implementation of {@link io.opencensus.metrics.DoubleGaugeMetric.Point}. */
+  public static final class PointImpl<T> extends Point {
+    private final AtomicDouble value = new AtomicDouble(0);
```

Comment:
I would do a bit different here:
1. Add an interface in the impl_core for the moment called Meter:
```
interface Meter {
  Metric getMetric();
}
```
2. The map in the Gauge should be <List<values>, Meter>
3. PointImpl extends Point implements Meter
4. Add a class PointWithFunction implements Meter.

I think we complicate things here a bit. By saying that people can retrieve the Point, what do we do if first add with function then add without function? Maybe consider to not let people retrieve the same Point and return an error if already exists?

Reply:
I have already added same interface (```Meter```) but it is implemented by both LongGaugeMetric and DoubleGaugeMetric. With this it is convenient to register Meters in MetricRegistryImpl. If we go with your approach, does it mean we are creating Metric for each individual points? Let me know if I am missing something.

Reply:
So my point here is that in this class you duplicate the logic of a point and a obj, function. You should have a common interface that returns a TimeSeries:

```java
interface XYZ {
  TimeSeries getTimeSeries();
}

// for AddPoint without function.
PointImpl extends Point implements XYZ {
}

// For when use with AddPoint with function
ABC implements XYZ {
}
```

Reply:
Also one more point that I made was to throw an exception if people try to create a Point/TimeSeries once with a function then without a function.

Also reading more the code seems that the function should be "createTimeSeries" or "addTimeSeries" because they add/create a TimeSeries and they get a Point back which is part of the TimeSeries.

Reply:
Changed the function name from addPoint to addTimeSeries.